frontend/src/components/phones/onboarding/RelayNumberPicker.tsx

```diff
@@ -82,29 +84,76 @@ const RelayNumberIntro = (props: RelayNumberIntroProps) => {
 
 type RelayNumberSelectionProps = {
   registerRelayNumber: (phoneNumber: string) => Promise<Response>;
+  search: (search: string) => Promise<Response | undefined>;
 };
 const RelayNumberSelection = (props: RelayNumberSelectionProps) => {
   const { l10n } = useLocalization();
   const relayNumberSuggestionsData = useRelayNumberSuggestions();
   const [phoneNumber, setPhoneNumber] = useState("");
+  const [searchValue, setSearchValue] = useState("");
   const [relayNumberIndex, setRelayNumberIndex] = useState(0);
-  const relayNumberSuggestions = relayNumberSuggestionsData.data
-    ? [
+  const [isSearching, setIsSearching] = useState(false);
+  const [relayNumberSuggestions, setRelayNumberSuggestions] = useState<
+    string[]
+  >([]);
+
+  if (relayNumberSuggestionsData.data && relayNumberSuggestions.length === 0) {
+    setRelayNumberSuggestions(
+      [
         ...relayNumberSuggestionsData.data.same_area_options,
         ...relayNumberSuggestionsData.data.other_areas_options,
         ...relayNumberSuggestionsData.data.same_prefix_options,
       ].map((suggestion) => suggestion.phone_number)
-    : [];
+    );
+  }
 
-  const onChange: ChangeEventHandler<HTMLInputElement> = (event) => {
+  const onRelayNumberChange: ChangeEventHandler<HTMLInputElement> = (event) => {
     setPhoneNumber(event.target.value);
   };
 
-  const onSubmit: FormEventHandler = (event) => {
+  const onSearchChange: ChangeEventHandler<HTMLInputElement> = (event) => {
+    setSearchValue(event.target.value);
+  };
+
+  const onSubmitRelayNumber: FormEventHandler = (event) => {
     event.preventDefault();
     props.registerRelayNumber(phoneNumber);
   };
 
+  const onSubmitSearch: FormEventHandler = (event) => {
+    event.preventDefault();
+
+    // return early if search is currently taking place
+    if (isSearching) return;
+
+    // set search to true as we begin searching
+    setIsSearching(true);
+
+    // submit string for search to our API
+    props
+      .search(searchValue)
+      .then((res) => res?.json())
+      .then((data) => {
+        if (data?.length !== 0) {
+          // empty out current relay number suggestions
+          relayNumberSuggestions.length = 0;
```

Comment:
Blocking question: Woah, TIL that you can write to `Array.length`! That said, generally you don't want to directly write to state variables, but instead use the setter (`setRelayNumberSuggestions`) - otherwise the component won't re-render with the updated state. Is there a reason you're doing this? Feels like it should be directly overwritten by the next line. Maybe you just need a ` ?? []` after the `.map` to ensure that at least the current suggestions get removed if no suggestions were found?

Reply:
Yeah, that's a remnant of something else I was trying - I was making it so that I could keep using a `const` within an array. I wanted to see if I could still mutate the value and you can definitely clear the array and keep pushing new values, lol. But I just removed that line and implemented all the other feedback. I'm still not sure what approach we are taking with the actual data for the drop-down. For now, I have a .json file. I'll implement it and see if we have a different source we want to use.